File: web/app/ascii.js
/* Reader for ASCII FBX, producing the same tree shape as the WebAssembly
 * binary reader so that analysis and rendering do not care which one ran.
 *
 * ASCII files are text, and the work is tokenising rather than number
 * crunching, so this stays in JavaScript; the geometry it produces still goes
 * through the same WebAssembly triangulator.
 */
'use strict';

const FbxAscii = (function () {
  const VERSION_COMMENT = /;\s*FBX\s+(\d+)\.(\d+)\.(\d+)/i;
  const NODE_LINE = /^[ \t]*[A-Za-z_]\w*[ \t]*:/m;
  const INT_RE = /^[+-]?\d+$/;
  const FLOAT_RE = /^[+-]?(\d+\.\d*|\.\d+|\d+)([eE][+-]?\d+)?$/;
  const COUNT_RE = /^\*(\d+)$/;

  const SPECIAL_FLOATS = {
    '1.#inf': Infinity, '-1.#inf': -Infinity,
    '1.#ind': NaN, '-1.#ind': NaN, '1.#qnan': NaN, '-1.#qnan': NaN,
    nan: NaN, inf: Infinity, '-inf': -Infinity,
  };

  const DELIMITERS = new Set([' ', '\t', '\r', '\f', '\v', '\n', ';', '"', ':', ',', '{', '}']);

  function looksLikeAscii(text) {
    if (VERSION_COMMENT.test(text)) return true;
    const head = text.slice(0, 8192);
    if (head.includes('FBXHeaderExtension') || head.includes('FBXVersion')) return true;
    return (head.match(NODE_LINE) || []).length >= 2 && head.includes('{');
  }

  function tokenize(text) {
    const tokens = [];
    let pos = 0;
    let line = 1;
    const length = text.length;
    while (pos < length) {
      const ch = text[pos];
      if (ch === '\n') { tokens.push({ kind: 'nl', line }); line++; pos++; }
      else if (ch === ' ' || ch === '\t' || ch === '\r' || ch === '\f' || ch === '\v') pos++;
      else if (ch === ';') { const end = text.indexOf('\n', pos); pos = end < 0 ? length : end; }
      else if (ch === '"') {
        const end = text.indexOf('"', pos + 1);
        if (end < 0) throw new Error(`unterminated string on line ${line}`);
        let raw = text.slice(pos + 1, end);
        if (raw.includes('&quot;')) raw = raw.split('&quot;').join('"');
        tokens.push({ kind: 'str', text: raw, line });
        pos = end + 1;
      } else if (ch === ':') { tokens.push({ kind: ':', line }); pos++; }
      else if (ch === ',') { tokens.push({ kind: ',', line }); pos++; }
      else if (ch === '{') { tokens.push({ kind: '{', line }); pos++; }
      else if (ch === '}') { tokens.push({ kind: '}', line }); pos++; }
      else {
        const start = pos;
        while (pos < length && !DELIMITERS.has(text[pos])) pos++;
        tokens.push({ kind: 'word', text: text.slice(start, pos), line });
      }
    }
    tokens.push({ kind: 'eof', line });
    return tokens;
  }

  function parseValue(token) {
    if (token.kind === 'str') return { kind: 'string', value: token.text };
    const text = token.text;
    const count = COUNT_RE.exec(text);
    if (count) return { kind: 'count', value: parseInt(count[1], 10) };
    if (INT_RE.test(text)) return { kind: 'int', value: parseInt(text, 10) };
    if (FLOAT_RE.test(text)) return { kind: 'float', value: parseFloat(text) };
    const lowered = text.toLowerCase();
    if (lowered in SPECIAL_FLOATS) return { kind: 'float', value: SPECIAL_FLOATS[lowered] };
    return { kind: 'word', value: text };
  }

  function toProp(entry) {
    switch (entry.kind) {
      case 'count': return { code: '*', typeName: 'array-size', value: entry.value };
      case 'word': return { code: 'W', typeName: 'word', value: entry.value };
      case 'int': return { code: 'L', typeName: 'int64', value: entry.value };
      case 'float': return { code: 'D', typeName: 'float64', value: entry.value };
      case 'string': return { code: 'S', typeName: 'string', value: entry.value };
      default: return { code: 'W', typeName: 'word', value: String(entry.value) };
    }
  }

  /** Fold the values of an `a:` record into a single array property. */
  function foldArray(entries) {
    if (!entries.length) return [];
    const allInt = entries.every((e) => e.kind === 'int' || e.kind === 'count');
    const code = allInt ? 'l' : 'd';
    const values = entries.map((e) => (typeof e.value === 'number' ? e.value : 0));
    return [{
      code,
      typeName: (allInt ? 'int64' : 'float64') + '[]',
      array: { length: values.length, encoding: 0, byteLength: 0, dataOffset: 0 },
      // ASCII arrays are already decoded, so the values ride along.
      values,
      value: null,
    }];
  }

  function parse(text) {
    if (text.charCodeAt(0) === 0xfeff) text = text.slice(1);
    const tokens = tokenize(text);
    const warnings = [];
    let index = 0;

    const peek = (ahead = 0) => tokens[Math.min(index + ahead, tokens.length - 1)];
    const next = () => tokens[index++];
    const skipNewlines = () => { while (peek().kind === 'nl') index++; };

    function consumeSeparator() {
      if (peek().kind === ',') { index++; skipNewlines(); return true; }
      let ahead = 0;
      while (peek(ahead).kind === 'nl') ahead++;
      if (ahead && peek(ahead).kind === ',') {
        index += ahead + 1;
        skipNewlines();
        return true;
      }
      return false;
    }

    function parseProperties(name) {
      const entries = [];
      for (;;) {
        const token = peek();
        if (token.kind === 'nl' || token.kind === '{' || token.kind === '}' || token.kind === 'eof') break;
        if (token.kind === ',') { index++; entries.push({ kind: 'word', value: '' }); continue; }
        entries.push(parseValue(next()));
        if (!consumeSeparator()) break;
      }
      return name === 'a' ? foldArray(entries) : entries.map(toProp);
    }

    function parseRecords(depth, insideBraces) {
      const records = [];
      for (;;) {
        skipNewlines();
        const token = peek();
        if (token.kind === 'eof') {
          if (insideBraces) warnings.push(`file ends with an unclosed '{' (line ${token.line})`);
          return records;
        }
        if (token.kind === '}') {
          if (!insideBraces) { warnings.push(`stray '}' on line ${token.line}`); index++; continue; }
          return records;
        }
        const head = next();
        if (head.kind !== 'word' && head.kind !== 'str') {
          warnings.push(`unexpected token on line ${head.line}; skipped`);
          continue;
        }
        if (peek().kind !== ':') {
          warnings.push(`record '${head.text}' on line ${head.line} is not followed by ':'; skipped`);
          continue;
        }
        index++; // ':'
        const node = { name: head.text, props: parseProperties(head.text), children: [], line: head.line };
        if (peek().kind === '{') {
          index++;
          node.children = parseRecords(depth + 1, true);
          if (peek().kind === '}') index++;
        }
        records.push(node);
      }
    }

    const root = { name: '', props: [], children: parseRecords(0, false) };

    let version = null;
    let versionSource = null;
    const header = root.children.find((n) => n.name === 'FBXHeaderExtension');
    const declared = header && header.children.find((n) => n.name === 'FBXVersion');
    if (declared && declared.props.length && typeof declared.props[0].value === 'number') {
      version = declared.props[0].value;
      versionSource = 'FBXHeaderExtension';
    } else {
      const match = VERSION_COMMENT.exec(text.slice(0, 4096));
      if (match) {
        version = (+match[1]) * 1000 + (+match[2]) * 100 + (+match[3]);
        versionSource = 'header comment';
      } else {
        warnings.push('no FBXVersion record and no version comment; version unknown');
      }
    }
    if (!root.children.length) warnings.push('no records were found');

    return {
      encoding: 'ascii',
      version,
      versionSource,
      wideOffsets: false,
      hasFooter: false,
      footerVersion: null,
      fileSize: text.length,
      root,
      warnings,
    };
  }

  return { parse, looksLikeAscii };
})();

if (typeof module !== 'undefined' && module.exports) module.exports = FbxAscii;

File: fbxtool/ascii.py
"""Reader for the ASCII FBX format.

The ASCII form carries the same record tree as the binary form, written as::

    ; FBX 7.3.0 project file
    ; ----------------------

    FBXHeaderExtension:  {
        FBXVersion: 7300
        Creator: "FBX SDK/FBX Plugins version 2013.3"
    }
    Objects:  {
        Geometry: 140717632, "Geometry::", "Mesh" {
            Vertices: *24 {
                a: -1,-1,1,1,-1,1,1,1,1
            }
        }
    }

Records are ``Name: prop, prop {children}``; ``;`` starts a comment; a trailing
comma continues a property list onto the next line; and ``*N`` is the element
count that precedes a large array.  Since ASCII has no type codes, values are
typed by inspection and the ``a:`` records that hold bulk data are folded into
a single array property so that both readers describe arrays the same way.
"""

from __future__ import annotations

import re
from collections import deque
from typing import Iterator

from .model import ArrayInfo, Document, Node, ParseError, Property

__all__ = ["is_ascii_fbx", "parse_ascii"]

_MAX_DEPTH = 256

# Token kinds.
_NAME = "name"
_STRING = "string"
_WORD = "word"
_COLON = "colon"
_COMMA = "comma"
_LBRACE = "lbrace"
_RBRACE = "rbrace"
_NEWLINE = "newline"
_EOF = "eof"

_INT_RE = re.compile(r"^[+-]?\d+$")
_FLOAT_RE = re.compile(r"^[+-]?(\d+\.\d*|\.\d+|\d+)([eE][+-]?\d+)?$")
_COUNT_RE = re.compile(r"^\*(\d+)$")
_VERSION_COMMENT_RE = re.compile(r";\s*FBX\s+(\d+)\.(\d+)\.(\d+)", re.IGNORECASE)
_NODE_LINE_RE = re.compile(r"^\s*[A-Za-z_][\w]*\s*:", re.MULTILINE)

# IEEE specials as emitted by MSVC-based exporters.
_SPECIAL_FLOATS = {
    "1.#inf": float("inf"),
    "-1.#inf": float("-inf"),
    "1.#ind": float("nan"),
    "-1.#ind": float("nan"),
    "1.#qnan": float("nan"),
    "-1.#qnan": float("nan"),
    "nan": float("nan"),
    "inf": float("inf"),
    "-inf": float("-inf"),
}


class _Token:
    __slots__ = ("kind", "text", "line")

    def __init__(self, kind: str, text: str, line: int) -> None:
        self.kind = kind
        self.text = text
        self.line = line

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"<{self.kind} {self.text!r} line {self.line}>"


def is_ascii_fbx(text: str) -> bool:
    """Heuristic check that *text* is the head of an ASCII FBX file."""
    if _VERSION_COMMENT_RE.search(text):
        return True
    head = text[:8192]
    if "FBXHeaderExtension" in head or "FBXVersion" in head:
        return True
    # Fall back to shape: a couple of ``Name:`` records and a brace.
    return len(_NODE_LINE_RE.findall(head)) >= 2 and "{" in head


def parse_ascii(
    text: str,
    *,
    path: str | None = None,
    load_arrays: bool = False,
    max_array_values: int | None = None,
) -> Document:
    """Parse ASCII FBX *text* into a Document."""
    if text.startswith("﻿"):
        text = text[1:]

    doc = Document(
        root=Node(""),
        encoding="ascii",
        path=path,
        file_size=len(text.encode("utf-8", "replace")),
    )

    parser = _Parser(text, doc, load_arrays=load_arrays, max_array_values=max_array_values)
    doc.root.children.extend(parser.parse_records(depth=0, inside_braces=False))

    version = doc.root.path_value("FBXHeaderExtension", "FBXVersion")
    if isinstance(version, int):
        doc.version = version
        doc.version_source = "FBXHeaderExtension"
    else:
        match = _VERSION_COMMENT_RE.search(text[:4096])
        if match:
            major, minor, patch = (int(g) for g in match.groups())
            doc.version = major * 1000 + minor * 100 + patch
            doc.version_source = "header comment"
        else:
            doc.warn("no FBXVersion record and no version comment; version unknown")

    if not doc.root.children:
        doc.warn("no records were found")
    return doc


def _tokenize(text: str) -> Iterator[_Token]:
    length = len(text)
    pos = 0
    line = 1
    while pos < length:
        ch = text[pos]
        if ch == "\n":
            yield _Token(_NEWLINE, "\n", line)
            line += 1
            pos += 1
        elif ch in " \t\r\f\v":
            pos += 1
        elif ch == ";":
            end = text.find("\n", pos)
            pos = length if end < 0 else end
        elif ch == '"':
            start = pos + 1
            end = text.find('"', start)
            if end < 0:
                raise ParseError(f"unterminated string on line {line}")
            raw = text[start:end]
            yield _Token(_STRING, _unescape(raw), line)
            line += raw.count("\n")
            pos = end + 1
        elif ch == ":":
            yield _Token(_COLON, ":", line)
            pos += 1
        elif ch == ",":
            yield _Token(_COMMA, ",", line)
            pos += 1
        elif ch == "{":
            yield _Token(_LBRACE, "{", line)
            pos += 1
        elif ch == "}":
            yield _Token(_RBRACE, "}", line)
            pos += 1
        else:
            start = pos
            while pos < length and text[pos] not in ' \t\r\f\v\n;":,{}':
                pos += 1
            yield _Token(_WORD, text[start:pos], line)
    yield _Token(_EOF, "", line)


def _unescape(raw: str) -> str:
    if "&quot;" in raw:
        raw = raw.replace("&quot;", '"')
    if "\\" in raw:
        raw = raw.replace('\\"', '"')
    return raw


class _Parser:
    """Recursive-descent parser over a bounded-lookahead token stream."""

    def __init__(self, text: str, doc: Document, *, load_arrays: bool,
                 max_array_values: int | None) -> None:
        self._tokens = _tokenize(text)
        self._buffer: deque[_Token] = deque()
        self._doc = doc
        self._load_arrays = load_arrays
        self._max_array_values = max_array_values

    # -- token access -----------------------------------------------------
    def _peek(self, index: int = 0) -> _Token:
        while len(self._buffer) <= index:
            self._buffer.append(next(self._tokens))
        return self._buffer[index]

    def _next(self) -> _Token:
        token = self._peek()
        self._buffer.popleft()
        return token

    def _skip_newlines(self) -> None:
        while self._peek().kind == _NEWLINE:
            self._next()

    # -- grammar ----------------------------------------------------------
    def parse_records(self, depth: int, inside_braces: bool) -> list[Node]:
        if depth > _MAX_DEPTH:
            raise ParseError(f"record nesting deeper than {_MAX_DEPTH}")
        records: list[Node] = []
        while True:
            self._skip_newlines()
            token = self._peek()
            if token.kind == _EOF:
                if inside_braces:
                    self._doc.warn(f"file ends with an unclosed '{{' (line {token.line})")
                return records
            if token.kind == _RBRACE:
                if not inside_braces:
                    self._doc.warn(f"stray '}}' on line {token.line}")
                    self._next()
                    continue
                return records
            record = self._parse_record(depth)
            if record is not None:
                records.append(record)

    def _parse_record(self, depth: int) -> Node | None:
        token = self._next()
        if token.kind not in (_WORD, _STRING):
            self._doc.warn(f"unexpected {token.text!r} on line {token.line}; skipped")
            return None
        if self._peek().kind != _COLON:
            self._doc.warn(
                f"record {token.text!r} on line {token.line} is not followed by ':'; skipped"
            )
            return None
        self._next()  # consume ':'

        node = Node(name=token.text, line=token.line)
        node.props.extend(self._parse_properties(node.name))

        if self._peek().kind == _LBRACE:
            self._next()
            node.children.extend(self.parse_records(depth + 1, inside_braces=True))
            if self._peek().kind == _RBRACE:
                self._next()
        return node

    def _parse_properties(self, name: str) -> list[Property]:
        values: list[object] = []
        while True:
            token = self._peek()
            if token.kind in (_NEWLINE, _LBRACE, _RBRACE, _EOF):
                break
            if token.kind == _COMMA:  # empty slot, e.g. ``P: "a",,1``
                self._next()
                values.append(None)
                continue
            values.append(_parse_value(self._next()))
            if not self._consume_separator():
                break
        if name == "a":
            return _fold_array(values, self._load_arrays, self._max_array_values)
        return [_make_property(value) for value in values]

    def _consume_separator(self) -> bool:
        """Consume a ``,`` separator, including one wrapped across newlines."""
        if self._peek().kind == _COMMA:
            self._next()
            self._skip_newlines()
            return True
        # Some writers break the line *before* the comma.
        index = 0
        while self._peek(index).kind == _NEWLINE:
            index += 1
        if index and self._peek(index).kind == _COMMA:
            for _ in range(index + 1):
                self._next()
            self._skip_newlines()
            return True
        return False


def _parse_value(token: _Token):
    if token.kind == _STRING:
        return token.text
    text = token.text
    match = _COUNT_RE.match(text)
    if match:
        return _ArrayCount(int(match.group(1)))
    if _INT_RE.match(text):
        return int(text)
    if _FLOAT_RE.match(text):
        return float(text)
    lowered = text.lower()
    if lowered in _SPECIAL_FLOATS:
        return _SPECIAL_FLOATS[lowered]
    return _Word(text)


class _ArrayCount(int):
    """An ``*N`` array-size marker."""


class _Word(str):
    """An unquoted token that is not a number (e.g. ``T``, ``Y``)."""


def _make_property(value) -> Property:
    if isinstance(value, _ArrayCount):
        return Property("*", int(value))
    if isinstance(value, _Word):
        return Property("W", str(value))
    if isinstance(value, bool):
        return Property("C", value)
    if isinstance(value, int):
        return Property("L", value)
    if isinstance(value, float):
        return Property("D", value)
    if isinstance(value, str):
        return Property("S", value)
    return Property("W", "" if value is None else str(value))


def _fold_array(values: list, load_arrays: bool, max_array_values: int | None) -> list[Property]:
    """Collapse the values of an ``a:`` record into one array property."""
    if not values:
        return []
    code = "l" if all(isinstance(v, int) and not isinstance(v, bool) for v in values) else "d"
    info = ArrayInfo(length=len(values), encoding=0, byte_length=0)
    kept = None
    if load_arrays:
        kept = values if max_array_values is None else values[:max_array_values]
        kept = [float(v) if code == "d" else int(v) for v in kept if isinstance(v, (int, float))]
    return [Property(code, kept, info)]
